monai/deploy/operators/clara_viz_operator.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2022 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+import monai.deploy.core as md
+from monai.deploy.core import ExecutionContext, Image, InputContext, IOType, Operator, OutputContext
+from monai.deploy.utils.importutil import optional_import
+
+DataDefinition, _ = optional_import("clara.viz.core", name="DataDefinition")
+Widget, _ = optional_import("clara.viz.widgets", name="Widget")
+display, _ = optional_import("IPython.display", name="display")
+interactive, _ = optional_import("ipywidgets", name="interactive")
+Dropdown, _ = optional_import("ipywidgets", name="Dropdown")
+Box, _ = optional_import("ipywidgets", name="Box")
+VBox, _ = optional_import("ipywidgets", name="VBox")
+
+
+@md.input("image", Image, IOType.IN_MEMORY)
+@md.input("seg_image", Image, IOType.IN_MEMORY)
+@md.env(pip_packages=["clara.viz.core", "clara.viz.widgets", "IPython"])
+class ClaraVizOperator(Operator):
+    """
+    This operator uses Clara Viz to provide interactive view of a 3D volume including segmentation mask.
+    """
+
+    def __init__(self):
+        """Constructor of the operator."""
+        super().__init__()
+
+    @staticmethod
+    def _build_array(image, order):
+        numpy_array = image.asnumpy().copy()
```

Comment:
@MMelQin @gigony I have a strange bug here. If I don't copy the segmentation numpy array data here then the data is wrong. Clara Viz copies the data when the Widget set_datafinition() function is called below in line 101.
Not sure how this could happen. Python is ref-counting objects, it should not be possible that the data is overwritten between here and line 101.

Reply:
@AndreasHeumann you have actually brought a very important point, in general with regard to the App SDK Image object state changes.

In your code, does the DataDefinition.Array simply refer to the passed in numpy_array, and modify it within  DataDefinition.Array.flip_axes? If yes, then I'd think the original ndarray would be modified.

On to the App SDK [Image](https://github.com/Project-MONAI/monai-deploy-app-sdk/blob/main/monai/deploy/core/domain/image.py) class itself, which at present is a very primitive wrapper/accessor. Its asnumpy() function simply returns the _data object, which is ArrayLike and set via the Image constructor. Modifying the object returned asnumpy() would directly modify the encapsulated object.

In the inference application, the DICOMSeriesToVolumeOperator create the (density) Image object (type int16), the Inference operator's InMemImageReader get the object via asnumpy() and exposes it via its get_data() without copying, and the image pre-processing before inference directly transforms the data from calling get_data(), so the (density) image may get modified. It is not an issue (yet) for the inference app as the original (density) image is not needed once inference is done, but may cause issues for the ClaraViz operator instances, which may try to get to the input (density) image at different stage of the app execution.

We can document this behavior and require the accessor of Image.asnumpy() not to modify the return, or make a copy to modify, but the best would be to make the function(s) on the Image class more explicit.

Reply:
Clara Viz is not modifying the data, it's just copying it. Density is fine, it's just the segmentation data which is wrong. Maybe when doing a copy() the resulting memory is continuous, internally Clara Viz is always using continuous memory and I have a copy function from non-continuous Python data to the continuous internal memory. Maybe there is a bug in that code.

Unfortunately there is no concept of immutable or constant objects in Python, so documenting the accessor is the only thing we can do.

Reply:
NumPy copy is not a deep copy, maybe `copy.deepcopy` will work better if needed?

Reply:
You are right. But using the copy() already makes it go down the continuous path. And I just found the bug in the non-continuous copy function, everything now works without making a copy of the array.

I also found this https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html?highlight=writeable. We can make numpy array data immutable, just tried it out and it works:

```
>>> a = numpy.zeros(100)
>>> a[0] = 12
>>> a.flags.writeable = False
>>> a[0] = 12
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: assignment destination is read-only
>>> a.flags.writeable = True
>>> a[0] = 12
>>>
```

Reply:
@AndreasHeumann There is already an open issue https://github.com/Project-MONAI/monai-deploy-app-sdk/issues/171 to enhance the Image class. Properly managing the state of its encapsulated data object (in most cases numpy array) should be part of it. Thanks.